File: app/src/main/java/com/riders/thelab/ui/tabs/WorkingTabsActivity.java
package com.riders.thelab.ui.tabs;

import android.annotation.SuppressLint;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;
import com.riders.thelab.R;
import com.riders.thelab.ui.tabs.fragment.OneFragment;
import com.riders.thelab.ui.tabs.fragment.ThreeFragment;
import com.riders.thelab.ui.tabs.fragment.TwoFragment;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Michaël on 07/03/2017.
 */

@SuppressLint("NonConstantResourceId")
public class WorkingTabsActivity extends AppCompatActivity {

    @BindView(R.id.tabToolbar)
    Toolbar mTabToolbar;
    @BindView(R.id.tabViewPager)
    ViewPager mViewPager;
    @BindView(R.id.tabs)
    TabLayout mTabLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tabs);

        ButterKnife.bind(this);

        setSupportActionBar(mTabToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(getString(R.string.activity_title_tabs));

        setupViewPager(mViewPager);

        mTabLayout.setupWithViewPager(mViewPager);
    }

    private void setupViewPager(ViewPager viewPager) {
        ViewPagerAdapter adapter = new ViewPagerAdapter(getSupportFragmentManager());

        adapter.addFragment(new OneFragment(), "ONE");
        adapter.addFragment(new TwoFragment(), "TWO");
        adapter.addFragment(new ThreeFragment(), "THREE");

        viewPager.setAdapter(adapter);
    }


    private class ViewPagerAdapter extends FragmentPagerAdapter {

        private final List<Fragment> mFragmentList = new ArrayList<>();
        private final List<String> mFragmentTitleList = new ArrayList<>();

        public ViewPagerAdapter(FragmentManager manager) {
            super(manager);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragmentList.get(position);
        }

        @Override
        public int getCount() {
            return mFragmentList.size();
        }

        public void addFragment(Fragment fragment, String title) {
            mFragmentList.add(fragment);
            mFragmentTitleList.add(title);
        }

        @Override
        public CharSequence getPageTitle(int position) {

            // return null to display only the icon (if there's icons)
            return mFragmentTitleList.get(position);
        }
    }
}

File: app/src/main/java/com/riders/thelab/ui/tabs/WorkingTabsActivity.kt
package com.riders.thelab.ui.tabs

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayoutMediator
import com.riders.thelab.databinding.ActivityTabsBinding
import com.riders.thelab.ui.tabs.fragment.OneFragment
import com.riders.thelab.ui.tabs.fragment.ThreeFragment
import com.riders.thelab.ui.tabs.fragment.TwoFragment
import timber.log.Timber

class WorkingTabsActivity : AppCompatActivity() {

    private var _viewBinding: ActivityTabsBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _viewBinding!!


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _viewBinding = ActivityTabsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.tabToolbar)

        setupViewPager(binding.tabViewPager2)

        TabLayoutMediator(binding.tabs, binding.tabViewPager2) { tab, position ->
            tab.text = "OBJECT ${(position + 1)}"
        }.attach()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
            }
        }
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        Timber.e("onDestroy()")
        _viewBinding = null
    }


    private fun setupViewPager(viewPager: ViewPager2) {
        val adapter = WorkingTabsAdapter(supportFragmentManager, lifecycle)
        adapter.addFragment(OneFragment.newInstance(), "ONE")
        adapter.addFragment(TwoFragment.newInstance(), "TWO")
        adapter.addFragment(ThreeFragment.newInstance(), "THREE")
        viewPager.adapter = adapter
    }
}